etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,388 @@
+
+            // Find a non-Shibboleth user account with an email that matches the EPPN
+            _getUserMatches(nonShibUsers, userWithEppn, function(err, userMatch) {
+                if (err) {
+                    log().warn(err);
+                }
+
+                // Link the Shibboleth account to the old user account
+                _swapUserLogin(userMatch, userWithEppn, function(err) {
+                    if (err) {
+                        _writeErrorRow(userWithEppn, userMatch, 'Failed to link Shibboleth login to existing user');
+                    } else {
+                        mappedUsers.push(userMatch);
+                    }
+
+                    if ((shibbolethUsers.length -1) === shibbolethUsers.indexOf(userHash)) {
+                        _.each(_.difference(nonShibUsers, mappedUsers), function(user) {
+                            _writeErrorRow(null, user, util.format('No matching Shibboleth user found for user %s', user.email));
+                        });
+
+                        callback(null, mappedUsers);
+                    }
+                });
+            });
+        });
+    });
+};
+
+/**
+ * Write errors to a CSV file
+ *
+ * @param  {Object}     shibUser                The Shibboleth user involved
+ * @param  {Object}     oldUser                 The existing user we want to map to
+ * @param  {String}     message                 A short message detailing the issue
+ * @api private
+ */
+var _writeErrorRow = function(shibUser, oldUser, message) {
+    csvStream.write({
+        'shib_principal_id': shibUser && shibUser.principalId ? shibUser.principalId : '',
+        'shib_email': shibUser && shibUser.email ? shibUser.email : '',
+        'shib_EPPN': shibUser && shibUser.eppn ? shibUser.eppn : '',
+        'old_principal_id': oldUser && oldUser.principalId ? oldUser.principalId : '',
+        'old_email': oldUser && oldUser.email ? oldUser.email : '',
+        'message': message ? message : ''
+    });
+    errors++;
+};
+
+/**
+ * Find the Shibboleth EPPN of the user
+ *
+ * @param  {String}     loginId                 The loginId for the user
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {String}     callback.eppn           The Shibboleth EPPN
+ * @api private
+ */
+function _findShibbolethEppn(loginId, callback) {
+    Cassandra.runQuery('SELECT "allAttributes" FROM "ShibbolethMetadata" WHERE "loginId" = ?', [loginId], function(err, rows) {
+        if (err) {
+            log().error({'err': err}, 'Failed Cassandra query for Shibboleth metadata');
+            return callback(err);
+        }
+
+        // Get the EPPN value from the Shibboleth metadata (stored as JSON)
+        var eppn = _.propertyOf(JSON.parse(rows[0].get('allAttributes').value))('eppn');
+
+        return callback(null, eppn);
+    });
+}
```

Comment:
In the migration case where users haven't authenticated with shibboleth yet, we won't have any shibboleth metadata available. Thankfully, the shibboleth `eppn` is set to the email address of the user, so we can copy over the existing gmail address associated to their gmail authentication login id entry as their equivalent shibboleth login id.